sql: add safety gates for pausable portal cleanup and flow invalidation

ZhouXing19 opened the pull request:
Informs: https://github.com/cockroachlabs/support/issues/3463

When a pausable portal encounters an error during execution, two issues can lead to panics on subsequent resume attempts:

1. The underlying FlowBase gets reset to nil during cleanup, but the portal's flow reference remains non-nil, causing hasFlowForPausablePortal() to incorrectly return true.

2. Errored portals are not removed from the portal map because deletion only occurs when execStmt() returns a non-nil fsm.Event.

This change adds two defensive checks:
- Nil the whole flow object hanging off the `portalInfo` while cleaning up the flow.
- Ensure errored portals are properly cleaned up regardless of event state

These gates prevent nil pointer dereferences when resuming portals that have been partially cleaned up due to errors.

Release note: None

yuzefovich reviewed (commented):
> @yuzefovich reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155669)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`-- commits` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155669#-Oc6oOdR7x4UxBAYLoER:-Oc6oOdR7x4UxBAYLoES:b-zhrp3j):*
nit: I'd probably omit a release note here given that we cannot explain well to users which conditions were necessary for the bug to happen.
___
*[`pkg/sql/conn_executor_exec.go` line 1201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155669#-Oc6nfEYAAOQbKLS0u4n:-Oc6nfEYAAOQbKLS0u4o:b-3igggm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3013d0a46bf5d7f5d1aec0f8784e7d2d33503d98/pkg/sql/conn_executor_exec.go#L1201)):*
> ```Go
> 			updateRetErrAndPayload(retErr, retPayload)
> 			portal.pauseInfo.resumableFlow.cleanup.run(ctx)
> 			portal.pauseInfo.resumableFlow.flow = nil
> ```

nit: rather than unsetting `flow` after each time `cleanup` runs, we should modify the cleanup function itself, this will be more bullet-proof.


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql: add safety gates for pausable portal cleanup and flow invalidation

Epic: None
Informs: cockroachlabs/support#3463

When a pausable portal encounters an error during execution, two issues
can lead to panics on subsequent resume attempts:

1. The underlying FlowBase gets reset to nil during cleanup, but the
 portal's flow reference remains non-nil, causing hasFlowForPausablePortal()
 to incorrectly return true.

2. Errored portals are not removed from the portal map because deletion
 only occurs when execStmt() returns a non-nil fsm.Event.

This change adds two defensive checks:
- Nil the whole flow object hanging off the portalInfo while cleaning up the flow.
- Ensure errored portals are properly cleaned up regardless of event state

These gates prevent nil pointer dereferences when resuming portals that
have been partially cleaned up due to errors.

Release note: None

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155669)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`-- commits` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155669#-Oc6oOdR7x4UxBAYLoER:-Oc7cTpd6RnIDEGFhKeg:b-896fix):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I'd probably omit a release note here given that we cannot explain well to users which conditions were necessary for the bug to happen.
</blockquote></details>

Done.
___
*[`pkg/sql/conn_executor_exec.go` line 1201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155669#-Oc6nfEYAAOQbKLS0u4n:-Oc7cUvh7dfbxSuo_2bq:bz5py0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3013d0a46bf5d7f5d1aec0f8784e7d2d33503d98/pkg/sql/conn_executor_exec.go#L1201)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: rather than unsetting `flow` after each time `cleanup` runs, we should modify the cleanup function itself, this will be more bullet-proof.
</blockquote></details>

Good idea! Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I think it's worth backporting to 25.3 and 25.4, so I added the labels. Also make sure to update the PR description.

> @yuzefovich reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155669#-:-Oc8DfWV6MLVYiiJwQJZ:bg1gyl2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @ZhouXing19)



<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
bors r+

michae2 commented:
blathers backport 25.2
